### PR TITLE
[RN] Support links in chat

### PR DESCRIPTION
**Pull request by zbettenbuk**

### CODE REVIEW

File: react/features/chat/Tokenizer.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    static tokenize(message: string): Array<Token> {
+        // Construct an initial array of tokens with a single entry, that we'll feed the tokenizer chain with.
+        let tokenizedArray = [
+            new Token(message)
+        ];
+
+        const SUPPORTED_TOKENIZERS = [
+
+            // Link tokenizer
+            new Tokenizer('link', 'https?://(?:[^., ]|[.,](?=\\w))+')
```
Comment (virtuacoplenny) [inline react/features/chat/Tokenizer.js]: Web used to have this approach too with links and emoji but then a library was found that does it, and it ended up fixing some bugs too. Is there no sufficient library for react-native?

If you really want your own, would you want to loosen up this regex not to include https?

File: react/features/chat/Token.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * Replresents a message token, that has a type based on what our app supports.
+ */
+export default class Token {
+    type: string;
+    content: string;
+
+    /**
+     * Constructs a new instance of the token based on the message part it receives.
+     *
+     * @param {string} messagePart - A message part (token) of the priginal message.
```
Comment (virtuacoplenny) [inline react/features/chat/Token.js]: priginal

File: react/features/chat/Tokenizer.js (hunk 1 above)
Comment (damencho) [reply react/features/chat/Tokenizer.js]: There were few bugs that got fixed by library ... I cannot recall the scenarios though :(

**Comment by zbettenbuk** [comment]
@saghul @virtuacoplenny @damencho I'm a big fan of 3rd party libs but in this came I would like to push this own implementation through for the following reasons:

- Unlike using multiple 3rd party libraries It's a less than 100 lines of easy to understand code that is going to handle all cases, like links, emojis, highlights (also: consider IP!)
- We've burned ourselves with 3rd party libraries a lot of times in the past, just remember the chat lib that we removed recently for the same reason, the calendar lib that we're continuously patching and some others that I can't even recall
- This is a logic-only implementation, meaning both web and mobile can (and should) use it, making it a consistent chat experience
- This only relies on regexp which provides great performance and can be extended easily without doing actual JS coding, or contributing into 3rd party libs

About the other points, like typos and the link component, I'll address them, of course.

**Comment by damencho** [comment]
Well, we had tried that in the past for the chat and it was broken all the time, new cases were coming up now and then, like security issues for not escaping, not detecting links, over detecting emojis where we were replacing part of links ... and going with both libs fix all the problems ... So I vote for sticking with the libraries.
One thing that we do not have while the libs have are a lot of users of that code, so many people had tested it and catch bugs and all corner cases, which we can catch and fix but again we need to run for long time and fix them one by one as they come ...

**Comment by saghul** [comment]
@zbettenbuk I appreciate the effort (I really do), but we have been there (hand rolled tokenization) and have improved since by using the mentioned libraries, while fixing a security issue in between.

Here is why I stand a strong -1:

>     * Unlike using multiple 3rd party libraries It's a less than 100 lines of easy to understand code that is going to handle all cases, like links, emojis, highlights (also: consider IP!)

It's not going to handle all cases. It's already subpar because it doesn't handle URLs without a schema. Once you add support for that and emoji you'll be way above 100 lines, so this point is moot. You cannot compare the size because the functionality is not the same.

>     * We've burned ourselves with 3rd party libraries a lot of times in the past, just remember the chat lib that we removed recently for the same reason, the calendar lib that we're continuously patching and some others that I can't even recall

We can't rewrite the universe and this is a great slippery slope for NIH-ing all sorts of things. Yes, we have needed to patch 3rd party librariers, but we wouldn't have been able to ship anything in a reasonable timeline without them.

>     * This is a logic-only implementation, meaning both web and mobile can (and should) use it, making it a consistent chat experience

Absolutely not. The chat implementation was updated to use the aforementioned libraries recently, so it's mobile who should be taking example from web, since web has been there and seen the dragons. https://github.com/jitsi/jitsi-meet/commit/f11b6cbb1e6311a2fc8520d0d3761538fd1f2385

>     * This only relies on regexp which provides great performance and can be extended easily without doing actual JS coding, or contributing into 3rd party libs

Unless you can prove your implementation provides the same funcionality with a significant performance improvement this argument is not sustained.

**Comment by zbettenbuk** [comment]
Well guys, I understand and appreciate your comments. It was a bold effort to get links quickly into chat, but since it seems so it has to be done again from scratch (with 3rd party libs) and due to the fact that it's not a priority now, I close the PR and get back to this when the task normally gets scheduled.

**Pull request closed without merging**